state/protocol/badger/state.go

```diff
@@ -622,18 +632,30 @@ func (state *State) AtHeight(height uint64) protocol.Snapshot {
 		// critical storage error
 		return invalid.NewSnapshotf("could not look up block by height: %w", err)
 	}
-	return state.AtBlockID(blockID)
+	return newSnapshotWithIncorporatedReferenceBlock(state, blockID)
 }
 
+// AtBlockID returns a snapshot for the block with the given ID. The block may be
+// finalized or un-finalized.
+// This function may return an invalid.Snapshot with:
+//   - state.ErrUnknownSnapshotReference:
+//     -> if no block with the given ID exists in the state
+//   - exception for critical unexpected storage errors
 func (state *State) AtBlockID(blockID flow.Identifier) protocol.Snapshot {
-	// TODO should return invalid.NewSnapshot(ErrUnknownSnapshotReference) if block doesn't exist
-	return NewSnapshot(state, blockID)
+	exists, err := state.headers.Exists(blockID)
+	if err != nil {
+		return invalid.NewSnapshotf("could not check existence of reference block")
```

Comment:
should this include the error message for context?

Reply:
https://github.com/onflow/flow-go/pull/4093/commits/b81799f88d0c3ca17632a40fd888728edead45c4